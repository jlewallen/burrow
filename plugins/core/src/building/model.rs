use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
