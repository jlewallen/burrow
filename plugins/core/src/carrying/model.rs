use crate::{library::model::*, tools};

pub use kernel::common::Carrying;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Containing {
    pub(crate) holding: Vec<EntityRef>,
    pub(crate) capacity: Option<u32>,
    pub(crate) produces: HashMap<String, String>,
}

impl Scope for Containing {
    fn scope_key() -> &'static str {
        "containing"
    }
}

impl Containing {
    pub fn start_carrying(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        if let Some(carryable) = item.scope::<Carryable>()? {
            let holding = self
                .holding
                .iter()
                .map(|h| h.to_entity())
                .collect::<Result<Vec<_>, _>>()?;

            for held in holding {
                if is_kind(&held, &carryable.kind)? {
                    let mut combining = held.scope_mut::<Carryable>()?;

                    combining.increase_quantity(&carryable.quantity.into())?;

                    combining.save()?;

                    get_my_session()?.obliterate(item)?;

                    return Ok(true);
                }
            }
        }

        if !self.is_holding(item) {
            self.holding.push(item.entity_ref());
        }

        Ok(true)
    }

    pub fn is_holding(&self, item: &EntityPtr) -> bool {
        self.holding.iter().any(|i| *i.key() == item.key())
    }

    fn remove_item(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        self.holding = self
            .holding
            .iter()
            .flat_map(|i| {
                if *i.key() == item.key() {
                    vec![]
                } else {
                    vec![i.clone()]
                }
            })
            .collect::<Vec<EntityRef>>()
            .to_vec();

        Ok(true)
    }

    pub fn stop_carrying(&mut self, found: Found) -> Result<Option<EntityPtr>, DomainError> {
        let item = found.entity()?;

        if !self.is_holding(item) {
            return Ok(None);
        }

        if let Some(carryable) = item.scope::<Carryable>()? {
            match &found {
                Found::One(item) => {
                    if carryable.quantity > 1.0 {
                        let (_original, separated) = tools::separate(item, &1.0.into())?;

                        Ok(Some(separated))
                    } else {
                        self.remove_item(item)?;

                        Ok(Some(item.clone()))
                    }
                }
                Found::Quantified(q, entity) => {
                    if carryable.quantity > q.as_f32() {
                        let (_original, separated) = tools::separate(entity, q)?;

                        Ok(Some(separated))
                    } else {
                        Ok(None)
                    }
                }
            }
        } else {
            self.remove_item(item)?;

            Ok(Some(item.clone()))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Carryable {
    kind: Kind,
    quantity: f32,
}

fn is_kind(entity: &EntityPtr, kind: &Kind) -> Result<bool, DomainError> {
    if let Some(carryable) = entity.scope::<Carryable>()? {
        Ok(*carryable.kind() == *kind)
    } else {
        Ok(false)
    }
}

impl Default for Carryable {
    fn default() -> Self {
        let session = get_my_session().expect("No session in Entity::new_blank!");
        Self {
            kind: Kind::new(session.new_identity()),
            quantity: 1.0,
        }
    }
}

impl Carryable {
    pub fn quantity(&self) -> f32 {
        self.quantity
    }

    pub fn decrease_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        if *q < 1.0.into() || *q > self.quantity.into() {
            Err(DomainError::Impossible)
        } else {
            self.quantity -= q.as_f32();

            Ok(self)
        }
    }

    pub fn increase_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        self.quantity += q.as_f32();

        Ok(self)
    }

    pub fn set_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.quantity = q.as_f32();

        Ok(self)
    }

    pub fn kind(&self) -> &Kind {
        &self.kind
    }

    pub fn set_kind(&mut self, kind: &Kind) {
        self.kind = kind.clone();
    }

    // Migrate items that were initialized with 0 quantities.
    fn sanity_check_quantity(&mut self) {
        if self.quantity < 1.0 {
            self.quantity = 1.0
        }
    }
}

impl Scope for Carryable {
    fn scope_key() -> &'static str {
        "carryable"
    }
}
