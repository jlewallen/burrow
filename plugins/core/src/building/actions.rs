use std::{rc::Rc, str::FromStr};

use chrono::Utc;

use crate::{
    building::model::{Constructed, QuickEdit},
    carrying::model::{Carryable, Containing},
    library::actions::*,
    looking::{actions::LookAction, model::new_area_observation},
    memory::model::{remember, EntityEvent, MemoryEvent},
    moving::model::Occupyable,
};

#[action]
pub struct AddScopeAction {
    pub scope_key: String,
}

impl Action for AddScopeAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        // TODO Security wise we would probably need a global list of who could
        // manipulate which scopes eventually.
        // TODO Right now this requires scopes to be functionable if all their
        // fields are ommitted. Look into `#[serde(default)]` to make this work
        // w/o a bunch of Option's?
        let Some(item) = tools::holding_one_item(surroundings.living())? else {
            return Ok(SimpleReply::NotFound.try_into()?);
        };

        debug!(item = ?item, scope_key = %self.scope_key, "add-scope");

        let mut item = item.entity().borrow_mut();

        item.add_scope_by_key(&self.scope_key);

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct EditAction {
    pub item: Item,
}

impl Action for EditAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let quick_edit: QuickEdit = (&editing).try_into()?;
                Ok(EditorReply::new(
                    editing.key().to_string(),
                    WorkingCopy::Markdown(quick_edit.to_string()),
                    SaveQuickEditAction::new_template(editing.key().clone())?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct EditRawAction {
    pub item: Item,
}

impl Action for EditRawAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let json = {
                    let editing = editing.borrow();
                    editing.to_json_value()?
                };
                let key = editing.key().clone();
                Ok(EditorReply::new(
                    key.to_string(),
                    WorkingCopy::Json(json),
                    SaveEntityJsonAction::new_template(key)?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DuplicateAction {
    pub item: Item,
}

impl Action for DuplicateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("duplicating {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(duplicating) => {
                info!("duplicating {:?}", duplicating);
                _ = tools::duplicate(&duplicating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct ObliterateAction {
    pub item: Item,
}

impl Action for ObliterateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("obliterate {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(obliterating) => {
                info!("obliterate {:?}", obliterating);
                tools::obliterate(&obliterating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct MakeItemAction {
    pub name: String,
}

impl Action for MakeItemAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("make-item {:?}", self.name);

        let creator = surroundings.living();

        let new_item: Entity = build_entity()
            .default_scope::<Carryable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_item = session.add_entity(new_item)?;

        tools::set_quantity(&new_item, 1f32)?;
        tools::set_container(creator, &vec![new_item.clone()])?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Created(EntityEvent {
                key: new_item.key().clone(),
                gid: new_item.gid(),
                name: new_item.name()?.unwrap(),
            }),
        )?;

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct BidirectionalDigAction {
    pub outgoing: String,
    pub returning: String,
    pub new_area: String,
}

impl Action for BidirectionalDigAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!(
            "bidirectional-dig {:?} <-> {:?} '{:?}'",
            self.outgoing, self.returning, self.new_area
        );

        let (_, living, area) = surroundings.unpack();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Occupyable>()?
            .default_scope::<Containing>()?
            .name(&self.new_area)
            .desc(&self.new_area)
            .try_into()?;
        let new_area = session.add_entity(new_area)?;

        tools::add_route(&area, &self.outgoing, &new_area)?;
        tools::add_route(&new_area, &self.returning, &area)?;

        match tools::navigate_between(&area, &new_area, &living)? {
            true => Ok(session.perform(Perform::Living {
                living,
                action: PerformAction::Instance(Rc::new(LookAction {})),
            })?),
            false => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveQuickEditAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveQuickEditAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Markdown(JSON_TEMPLATE_VALUE_SENTINEL.to_owned());
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveQuickEditAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:quick-edit {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Markdown(text) => {
                    let quick = QuickEdit::from_str(text)?;
                    let mut entity = entity.borrow_mut();
                    if let Some(name) = quick.name {
                        entity.set_name(&name)?;
                    }
                    if let Some(desc) = quick.desc {
                        entity.set_desc(&desc)?;
                    }

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveEntityJsonAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveEntityJsonAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Json(JsonValue::String(JSON_TEMPLATE_VALUE_SENTINEL.to_owned()));
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveEntityJsonAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:entity-json {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Json(value) => {
                    let replacing = Entity::from_value(value.clone())?;
                    entity.replace(replacing);

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct BuildAreaAction {
    pub name: String,
}

impl Action for BuildAreaAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("build-area {:?}", self.name);

        let creator = surroundings.living();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Containing>()?
            .default_scope::<Occupyable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_area = session.add_entity(new_area)?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Constructed(EntityEvent {
                key: new_area.key().clone(),
                gid: new_area.gid(),
                name: new_area.name()?.unwrap(),
            }),
        )?;

        info!("created {:?}", new_area);

        let observed = new_area_observation(creator, &new_area)?;
        let reply = Constructed::Area(observed);
        Ok(reply.try_into()?)
    }
}
