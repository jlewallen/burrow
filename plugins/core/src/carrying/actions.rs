use crate::{carrying::model::Carrying, library::actions::*, looking::model::Observe};

#[action]
pub struct HoldAction {
    pub item: Item,
}

impl Action for HoldAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("hold {:?}!", self.item);

        let (_, actor, area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(holding) => match tools::move_between(&area, &actor, holding.clone())? {
                true => Ok(reply_ok(
                    actor.clone(),
                    Audience::Area(area.key().clone()),
                    Carrying::Held {
                        actor: (&actor).observe(&actor)?.expect("No observed entity"),
                        item: (&holding.entity()?)
                            .observe(&actor)?
                            .expect("No observed entity"),
                        area: (&area).observe(&actor)?.expect("No observed entity"),
                    },
                )?),
                false => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DropAction {
    pub maybe_item: Option<Item>,
}

impl Action for DropAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("drop {:?}!", self.maybe_item);

        let (_, actor, area) = surroundings.unpack();

        match &self.maybe_item {
            Some(item) => match session.find_item(surroundings, item)? {
                Some(dropping) => match tools::move_between(&actor, &area, dropping.clone())? {
                    true => Ok(reply_ok(
                        actor.clone(),
                        Audience::Area(area.key().clone()),
                        Carrying::Dropped {
                            actor: (&actor).observe(&actor)?.expect("No observed entity"),
                            item: (&dropping.entity()?)
                                .observe(&actor)?
                                .expect("No observed entity"),
                            area: (&area).observe(&actor)?.expect("No observed entity"),
                        },
                    )?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct PutInsideAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for PutInsideAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("put-inside {:?} -> {:?}", self.item, self.vessel);

        let (_, _user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.vessel)? {
                Some(vessel) => {
                    let vessel = vessel.one()?;
                    if tools::is_container(&vessel)? {
                        let from = tools::container_of(&item.clone().one()?)?;
                        match tools::move_between(&from, &vessel, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        }
                    } else {
                        Ok(SimpleReply::Impossible.try_into()?)
                    }
                }
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct TakeOutAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for TakeOutAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("take-out {:?} -> {:?}", self.item, self.vessel);

        let (_, user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.vessel)? {
            Some(vessel) => {
                let vessel = vessel.one()?;
                if tools::is_container(&vessel)? {
                    match session.find_item(surroundings, &self.item)? {
                        Some(item) => match tools::move_between(&vessel, &user, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        },
                        None => Ok(SimpleReply::NotFound.try_into()?),
                    }
                } else {
                    Ok(SimpleReply::Impossible.try_into()?)
                }
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct GiveToAction {
    pub item: Item,
    pub receiver: Item,
}

impl Action for GiveToAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("give-to {:?} -> {:?}", self.item, self.receiver);

        let (_, user, _area) = surroundings.unpack();

        // I think there are very interesting permission related implications
        // here. For example, limiting third party access to your hands except
        // for key individuals.
        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.receiver)? {
                Some(receiver) => match tools::move_between(&user, &receiver.one()?, item)? {
                    true => Ok(SimpleReply::Done.try_into()?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}
