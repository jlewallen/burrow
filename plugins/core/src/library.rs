pub mod plugin {
    pub use anyhow::Result;
    pub use serde::Deserialize;
    pub use tracing::*;

    pub use super::parser::try_parsing;
    pub use kernel::common::DeserializeTagged;
    pub use kernel::prelude::*;
    pub use kernel::try_deserialize_all;
}

pub mod parser {
    use nom::sequence::delimited;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, opt, recognize},
        sequence::{pair, preceded, separated_pair, terminated, tuple},
        IResult,
    };
    pub use tracing::*;

    pub use kernel::prelude::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn text_to_end_of_line(i: &str) -> IResult<&str, &str> {
        take_while1(|_c: char| true)(i)
    }

    pub fn string_literal(i: &str) -> IResult<&str, &str> {
        delimited(tag("\""), string_inside, tag("\""))(i)
    }

    fn string_inside(i: &str) -> IResult<&str, &str> {
        take_while(|c: char| c.is_alphabetic() || c.is_whitespace())(i)
    }

    pub fn camel_case_word(i: &str) -> IResult<&str, &str> {
        word(i) // TODO
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::Gid(EntityGid::new(n))
        })(i)
    }

    pub fn surrounding_area(i: &str) -> IResult<&str, Item> {
        map(alt((tag("area"), tag("here"))), |_s: &str| Item::Area)(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((surrounding_area, myself, noun, gid_reference))(i)
    }

    pub fn myself(i: &str) -> IResult<&str, Item> {
        map(alt((tag("self"), tag("myself"))), |_s: &str| Item::Myself)(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }

    pub fn try_parsing<T: ParsesActions>(parser: T, i: &str) -> EvaluationResult {
        parser.try_parse_action(i)
    }
}

pub mod model {
    pub use anyhow::Result;
    pub use chrono::{DateTime, Utc};
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::json;
    pub use std::rc::Rc;
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::tools;
    pub use anyhow::Result;
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use tracing::*;

    pub fn reply_ok<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<Effect> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(Effect::Ok)
    }

    pub fn reply_done<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<SimpleReply> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(SimpleReply::Done)
    }
}

pub mod tests {
    pub use anyhow::Result;
    pub use chrono::TimeZone;
    pub use chrono::Utc;
    pub use serde::de::DeserializeOwned;
    pub use std::collections::HashSet;
    pub use std::rc::Rc;

    pub use crate::tools;
    pub use crate::{BuildSurroundings, QuickThing};
    pub use kernel::common::*;
    pub use kernel::here;
    pub use kernel::prelude::*;
    pub use tracing::*;

    pub use super::plugin::try_parsing;

    pub trait ToDebugJson {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error>;
    }

    impl ToDebugJson for Effect {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error> {
            serde_json::to_value(self)
        }
    }

    pub fn parse_and_perform<T: ParsesActions>(
        parser: T,
        line: &str,
    ) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action = try_parsing(parser, line)?;
        let action = action.unwrap();
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }

    pub fn perform_directly<T: Action>(action: T) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action: Rc<dyn Action> = Rc::new(action);
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }
}
