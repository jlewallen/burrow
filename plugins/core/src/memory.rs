use crate::library::plugin::*;

#[derive(Default)]
pub struct MemoryPluginFactory {}

impl PluginFactory for MemoryPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(MemoryPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct MemoryPlugin {}

impl Plugin for MemoryPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "memory"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn sources(&self) -> Vec<Box<dyn ActionSource>> {
        vec![Box::new(ActionSources::default())]
    }
}

impl ParsesActions for MemoryPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::RecallActionParser {}, i)
    }
}

#[derive(Default)]
pub struct ActionSources {}

impl ActionSource for ActionSources {
    fn try_deserialize_action(
        &self,
        _tagged: &TaggedJson,
    ) -> Result<Option<Box<dyn Action>>, serde_json::Error> {
        Ok(None)
    }
}

pub mod model {
    use crate::library::model::*;

    #[derive(Debug, Serialize, ToTaggedJson)]
    #[serde(rename_all = "camelCase")]
    pub struct RecalledMemory {
        pub time: DateTime<Utc>,
        pub key: EntityKey,
        pub gid: EntityGid,
        pub name: String,
    }

    impl From<SpecificMemory> for RecalledMemory {
        fn from(value: SpecificMemory) -> Self {
            let entity = match value.event {
                Memory::Created(e) => e,
                Memory::Destroyed(e) => e,
                Memory::Constructed(e) => e,
            };

            Self {
                time: value.time,
                key: entity.key,
                gid: entity.gid,
                name: entity.name,
            }
        }
    }

    #[derive(Debug, Serialize, ToTaggedJson)]
    #[serde(rename_all = "camelCase")]
    pub struct RecallReply {
        pub memories: Vec<RecalledMemory>,
    }

    impl Reply for RecallReply {}

    impl TryInto<Effect> for RecallReply {
        type Error = TaggedJsonError;

        fn try_into(self) -> std::result::Result<Effect, Self::Error> {
            Ok(Effect::Reply(EffectReply::TaggedJson(
                TaggedJson::new_from(serde_json::to_value(self)?)?,
            )))
        }
    }

    #[derive(Debug, Serialize, Deserialize, Clone)]
    pub struct EntityEvent {
        pub(crate) key: EntityKey,
        pub(crate) gid: EntityGid,
        pub(crate) name: String,
    }

    #[derive(Debug, Serialize, Deserialize, Clone)]
    pub enum Memory {
        Created(EntityEvent),
        Destroyed(EntityEvent),
        Constructed(EntityEvent),
    }

    #[derive(Debug, Serialize, Deserialize, Clone)]
    pub struct SpecificMemory {
        pub time: DateTime<Utc>,
        pub event: Memory,
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Mind {
        memory: Vec<SpecificMemory>,
    }

    impl Scope for Mind {
        fn scope_key() -> &'static str {
            "memory"
        }
    }

    impl From<Mind> for Vec<SpecificMemory> {
        fn from(value: Mind) -> Self {
            value.memory
        }
    }

    pub fn memories_of(entity: &EntityPtr) -> Result<Vec<SpecificMemory>, DomainError> {
        let memory = entity.scope::<Mind>()?.unwrap_or_default();
        Ok(memory.memory.clone())
    }

    pub fn remember(
        entity: &EntityPtr,
        time: DateTime<Utc>,
        event: Memory,
    ) -> Result<(), DomainError> {
        let mut memory = entity.scope_mut::<Mind>()?;
        memory.memory.push(SpecificMemory { time, event });
        memory.save()
    }
}

pub mod actions {
    use super::model::*;
    use crate::library::actions::*;

    #[action]
    pub struct RecallAction {}

    impl Action for RecallAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            let (_world, actor, _area) = surroundings.unpack();
            let memories = memories_of(&actor)?;
            Ok(RecallReply {
                memories: memories.into_iter().map(|m| m.into()).collect(),
            }
            .try_into()?)
        }
    }
}

pub mod parser {
    use super::actions::*;
    use crate::library::parser::*;

    pub struct RecallActionParser {}

    impl ParsesActions for RecallActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let (_, action) = map(tag("recall"), |_| {
                Box::new(RecallAction {}) as Box<dyn Action>
            })(i)?;

            Ok(Some(action))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::model::*;
    use super::parser::*;
    use crate::library::tests::*;

    #[test]
    fn it_recalls_when_no_memories() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().build()?;

        let action = try_parsing(RecallActionParser {}, "recall")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_debug_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_recalls_when_some_memories() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.build()?;

        let (_, actor, _) = surroundings.clone().unpack();
        let time = Utc.with_ymd_and_hms(1982, 4, 23, 0, 0, 0).unwrap();
        remember(
            &actor,
            time,
            Memory::Created(EntityEvent {
                key: session.new_key(),
                gid: EntityGid::new(3),
                name: "Doesn't actually exist".to_owned(),
            }),
        )?;

        let action = try_parsing(RecallActionParser {}, "recall")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, _person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_debug_json()?);

        build.close()?;

        Ok(())
    }
}
