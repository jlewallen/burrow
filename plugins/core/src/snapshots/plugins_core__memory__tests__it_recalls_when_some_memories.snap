---
source: plugins/core/src/memory.rs
expression: reply.to_debug_json()?
---
{
  "recallReply": {
    "memories": [
      {
        "gid": 3,
        "key": "E-3",
        "name": "Doesn't actually exist",
        "time": "1982-04-23T00:00:00Z"
      }
    ]
  }
}
